### ISSUE TITLE
won't work with dask

### ISSUE THREAD

**Issue by uqyge** [Cantera/cantera#678]
### Cantera version
2.4
### Operating System
ubuntu
### Python/MATLAB version
python
### Expected Behavior
parallel with dask
### Actual Behavior
TypeError: no default __reduce__ due to non-trivial __cinit__
### Steps to reproduce
1.
2.
3.

**Comment by bryanwweber** [comment]
Hi, can you please provide a complete script/steps to reproduce and the full error message in your post. Thank you.

**Comment by uqyge** [comment]
thanks for the quick response. I do it now.

**Comment by uqyge** [comment]
```python
import copy
import os
import shutil

import cantera as ct
import dask
import numpy as np

dask.config.set(scheduler='processes')


# Create directory for output data files
data_directory = 'diffusion_flame_batch_data/'
if os.path.exists(data_directory):
    shutil.rmtree(data_directory)
    os.makedirs(data_directory)

if not os.path.exists(data_directory):
    os.makedirs(data_directory)



class FlameExtinguished(Exception):
    pass


temperature_limit_extinction = 900  # K


def interrupt_extinction(t):
    if np.max(f.T) < temperature_limit_extinction:
        raise FlameExtinguished('Flame extinguished')
    return 0.

# PART 1: INITIALIZATION
reaction_mechanism = 'gri30.xml'
gas = ct.Solution(reaction_mechanism)
width = 0.02  # 18mm wide
grid_ini = width * np.linspace(0, 1, 20)

f = ct.CounterflowDiffusionFlame(gas, grid=grid_ini)
f.set_max_grid_points(f.flame, 10000)

# Define the operating pressure and boundary conditions
f.P = ct.one_atm  # 1 bar

f.fuel_inlet.mdot = 0.05  # kg/m^2/s

f.fuel_inlet.Y = 'CH4:1'
f.fuel_inlet.T = 300  # K

f.oxidizer_inlet.mdot = 0.1  # kg/m^2/s
f.oxidizer_inlet.Y = 'O2:0.23,N2:0.77'

f.oxidizer_inlet.T = 300  # K

# Set refinement parameters, if used
f.set_refine_criteria(ratio=100.0, slope=0.2, curve=0.5, prune=0.0001)

f.set_interrupt(interrupt_extinction)

# Initialize and solve
print('Creating the initial solution')
f.solve(loglevel=0, refine_grid=False, auto=True)

# Save to data directory
file_name = 'initial_solution.xml'
f.save(data_directory + file_name,
       name='solution',
       description='Cantera version ' + ct.__version__ +
       ', reaction mechanism ' + reaction_mechanism)

# STRAIN RATE LOOP
def` flamelet_gen(i):
    # Compute counterflow diffusion flames at increasing strain rates at 1 bar
    # The strain rate is assumed to increase by 25% in each step until the flame is
    # extinguished
    strain_factor = 1.1**i

    exp_d_a = -1. / 2.
    exp_u_a = 1. / 2.
    exp_V_a = 1.
    exp_lam_a = 2.
    exp_mdot_a = 1. / 2.

    # Restore initial solution
    file_name = 'initial_solution.xml'
    f.restore(filename=data_directory + file_name, name='solution', loglevel=0)

    # Do the strain rate loop
    if np.max(f.T) > temperature_limit_extinction:
        print('strain rate iteration', i)
        # Create an initial guess based on the previous solution
        # Update grid
        f.flame.grid *= strain_factor**exp_d_a
        normalized_grid = f.grid / (f.grid[-1] - f.grid[0])
        # Update mass fluxes
        f.fuel_inlet.mdot *= strain_factor**exp_mdot_a
        f.oxidizer_inlet.mdot *= strain_factor**exp_mdot_a
        # Update velocities
        f.set_profile('u', normalized_grid, f.u * strain_factor**exp_u_a)
        f.set_profile('V', normalized_grid, f.V * strain_factor**exp_V_a)
        # Update pressure curvature
        f.set_profile('lambda', normalized_grid,
                      f.L * strain_factor**exp_lam_a)
        try:
            # Try solving the flame
            f.solve(loglevel=0, refine_grid=True)
            file_name = 'strain_loop_' + format(i, '02d') + '.xml'
            f.save(data_directory + file_name,
                   name='solution',
                   loglevel=1,
                   description='Cantera version ' + ct.__version__ +
                   ', reaction mechanism ' + reaction_mechanism)
        except FlameExtinguished:
            print('Flame extinguished')

        except ct.CanteraError as e:
            print('Error occurred while solving:', e)

    print(f.strain_rate('max'))
    strain_max = f.strain_rate('max')
    return strain_max

# This is where it goes wrong.
a = [dask.delayed(flamelet_gen)(i) for i in range(4)]
dask.compute(*a)
```

**Comment by uqyge** [comment]
```python
￼
a = [dask.delayed(flamelet_gen)(i) for i in range(4)]...
￼
a = [dask.delayed(flamelet_gen)(i) for i in range(4)]...
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
 in 
      1 a = [dask.delayed(flamelet_gen)(i) for i in range(4)]
----> 2 dask.compute(*a)
      3 
      4 # import multiprocessing as mp
      5 

~/anaconda3/envs/tf2/lib/python3.7/site-packages/dask/base.py in compute(*args, **kwargs)
    396     keys = [x.__dask_keys__() for x in collections]
    397     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 398     results = schedule(dsk, keys, **kwargs)
    399     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    400 

~/anaconda3/envs/tf2/lib/python3.7/site-packages/dask/multiprocessing.py in get(dsk, keys, num_workers, func_loads, func_dumps, optimize_graph, pool, **kwargs)
    190                            get_id=_process_get_id, dumps=dumps, loads=loads,
    191                            pack_exception=pack_exception,
--> 192                            raise_exception=reraise, **kwargs)
    193     finally:
    194         if cleanup:

~/anaconda3/envs/tf2/lib/python3.7/site-packages/dask/local.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    447             # Seed initial tasks into the thread pool
    448             while state['ready'] and len(state['running']) < num_workers:
--> 449                 fire_task()
    450 
    451             # Main loop, wait on tasks to finish, insert new ones

~/anaconda3/envs/tf2/lib/python3.7/site-packages/dask/local.py in fire_task()
    441                 # Submit
    442                 apply_async(execute_task,
--> 443                             args=(key, dumps((dsk[key], data)),
    444                                   dumps, loads, get_id, pack_exception),
    445                             callback=queue.put)

~/anaconda3/envs/tf2/lib/python3.7/site-packages/dask/multiprocessing.py in _dumps(x)
     24 
     25 def _dumps(x):
---> 26     return cloudpickle.dumps(x, protocol=pickle.HIGHEST_PROTOCOL)
     27 
     28 

~/anaconda3/envs/tf2/lib/python3.7/site-packages/cloudpickle/cloudpickle.py in dumps(obj, protocol)
   1106     try:
   1107         cp = CloudPickler(file, protocol=protocol)
-> 1108         cp.dump(obj)
   1109         return file.getvalue()
   1110     finally:

~/anaconda3/envs/tf2/lib/python3.7/site-packages/cloudpickle/cloudpickle.py in dump(self, obj)
    471         self.inject_addons()
    472         try:
--> 473             return Pickler.dump(self, obj)
    474         except RuntimeError as e:
    475             if 'recursion' in e.args[0]:

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in dump(self, obj)
    435         if self.proto >= 4:
    436             self.framer.start_framing()
--> 437         self.save(obj)
    438         self.write(STOP)
    439         self.framer.end_framing()

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in save_tuple(self, obj)
    769         if n <= 3 and self.proto >= 2:
    770             for element in obj:
--> 771                 save(element)
    772             # Subtle.  Same as in the big comment below.
    773             if id(obj) in memo:

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in save_tuple(self, obj)
    769         if n <= 3 and self.proto >= 2:
    770             for element in obj:
--> 771                 save(element)
    772             # Subtle.  Same as in the big comment below.
    773             if id(obj) in memo:

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

~/anaconda3/envs/tf2/lib/python3.7/site-packages/cloudpickle/cloudpickle.py in save_function(self, obj, name)
    545             return self.save_pypy_builtin_func(obj)
    546         else:
--> 547             return self.save_function_tuple(obj)
    548 
    549     dispatch[types.FunctionType] = save_function

~/anaconda3/envs/tf2/lib/python3.7/site-packages/cloudpickle/cloudpickle.py in save_function_tuple(self, func)
    745         if hasattr(func, '__kwdefaults__'):
    746             state['kwdefaults'] = func.__kwdefaults__
--> 747         save(state)
    748         write(pickle.TUPLE)
    749         write(pickle.REDUCE)  # applies _fill_function on the tuple

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in save_dict(self, obj)
    854 
    855         self.memoize(obj)
--> 856         self._batch_setitems(obj.items())
    857 
    858     dispatch[dict] = save_dict

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in _batch_setitems(self, items)
    880                 for k, v in tmp:
    881                     save(k)
--> 882                     save(v)
    883                 write(SETITEMS)
    884             elif n:

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in save_dict(self, obj)
    854 
    855         self.memoize(obj)
--> 856         self._batch_setitems(obj.items())
    857 
    858     dispatch[dict] = save_dict

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in _batch_setitems(self, items)
    880                 for k, v in tmp:
    881                     save(k)
--> 882                     save(v)
    883                 write(SETITEMS)
    884             elif n:

~/anaconda3/envs/tf2/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    522             reduce = getattr(obj, "__reduce_ex__", None)
    523             if reduce is not None:
--> 524                 rv = reduce(self.proto)
    525             else:
    526                 reduce = getattr(obj, "__reduce__", None)

~/anaconda3/envs/tf2/lib/python3.7/site-packages/cantera/_cantera.cpython-37m-x86_64-linux-gnu.so in cantera._cantera.Sim1D.__reduce_cython__()

TypeError: no default __reduce__ due to non-trivial __cinit__
```

**Comment by ischoegl** [comment]
The call
```
~/anaconda3/envs/tf2/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
```
hints at the fact that this may fail as cantera objects are not picklable?

**Comment by uqyge** [comment]
I'm following the flamelet_batch.py tutorial and trying to create a parallel version with dask with no success. Right now I am using the multiprocessing package with python as a workaround, But I am limited to use only the cpu on my machine.

**Comment by ischoegl** [comment]
I haven't worked with dask, but my own tests with multiprocessing required the cantera objects to be initialized by workers. I.e. it fails where you start the loop
```
a = [dask.delayed(flamelet_gen)(i) for i in range(4)]
```
Try initializing the Sim1D objects within flamelet_gen instead ...

**Comment by ischoegl** [comment]
For further reading, see #570

**Comment by speth** [comment]
Indeed, there is no expectation that Cantera objects are pickleable, and even if they were, I suspect this would be an inefficient way of implementing parallelism with Cantera. I think the User's Group would be a better place to have a discussion of how to use Cantera in a parallel Python program.

**Comment by bryanwweber** [comment]
@uqyge Can you please ask this question on the Google User's Group (https://groups.google.com/forum/#!forum/cantera-users)? I also think this is a duplicate of #570 as @ischoegl mentioned, so I'll be closing it. Please feel free to reply if you disagree. Thanks!

**Comment by uqyge** [comment]
thanks for all the help and please close the issue.
